eventsources/sources/azureservicebus/start.go

```diff
@@ -0,0 +1,182 @@
+		return err
+	}
+
+	log.Info("connecting to the service bus...")
+	clientOptions := servicebus.ClientOptions{}
+	if servicebusEventSource.TLS != nil {
+		tlsConfig, err := common.GetTLSConfig(servicebusEventSource.TLS)
+		if err != nil {
+			log.Errorw("failed to get the tls configuration", zap.Error(err))
+			return err
+		}
+		clientOptions.TLSConfig = tlsConfig
+	}
+
+	client, err := servicebus.NewClientFromConnectionString(connStr, &clientOptions)
+	if err != nil {
+		log.Errorw("failed to create a service bus client", zap.Error(err))
+		return err
+	}
+
+	var receiver *servicebus.Receiver
+	var receiverType ReceiverType
+
+	if servicebusEventSource.QueueName != "" {
+		log.Info("creating a queue receiver...")
+		receiverType = ReceiverTypeQueue
+		receiver, err = client.NewReceiverForQueue(servicebusEventSource.QueueName, &servicebus.ReceiverOptions{
+			ReceiveMode: servicebus.ReceiveModeReceiveAndDelete,
+		})
+	} else {
+		log.Info("creating a subscription receiver...")
+		receiverType = ReceiverTypeSubscription
+		receiver, err = client.NewReceiverForSubscription(servicebusEventSource.TopicName, servicebusEventSource.SubscriptionName, &servicebus.ReceiverOptions{
+			ReceiveMode: servicebus.ReceiveModeReceiveAndDelete,
+		})
+	}
+	if err != nil {
+		if receiverType == ReceiverTypeQueue {
+			log.With("queue", servicebusEventSource.QueueName).Errorw("failed to create a queue receiver", zap.Error(err))
+		} else {
+			log.With("topic", servicebusEventSource.TopicName, "subscription", servicebusEventSource.SubscriptionName).Errorw("failed to create a receiver for subscription", zap.Error(err))
+		}
+		return err
+	}
+
+	if servicebusEventSource.JSONBody {
+		log.Info("assuming all events have a json body...")
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			log.Info("context has been cancelled, stopping the Azure Service Bus event source...")
+			if err := receiver.Close(ctx); err != nil {
+				log.Errorw("failed to close the receiver", zap.Error(err))
+				return err
+			}
+			return nil
+		default:
+			messages, err := receiver.ReceiveMessages(ctx, 1, nil)
+			if err != nil {
+				log.Errorw("failed to receive messages", zap.Error(err))
+				return err
+			}
+
+			for _, message := range messages {
+				if err := el.handleOne(servicebusEventSource, message, dispatch, log); err != nil {
+					if receiverType == ReceiverTypeQueue {
+						log.With("queue", servicebusEventSource.QueueName, "message_id", message.MessageID).Errorw("failed to process Azure Service Bus message", zap.Error(err))
+					} else {
+						log.With("topic", servicebusEventSource.TopicName, "subscription", servicebusEventSource.SubscriptionName, "message_id", message.MessageID).Errorw("failed to process Azure Service Bus message", zap.Error(err))
+					}
+					el.Metrics.EventProcessingFailed(el.GetEventSourceName(), el.GetEventName())
```

Comment:
Metrics for success is missing.

Reply:
Which metric is used for success? I tried to have a look at qmap or redisstream which only use `EventProcessingFailed` and `EventProcessingDuration`

Reply:
My bad, we don't have that one, `argo_events_events_sent_total` is used instead.